### PR TITLE
Record server startup time in ProfileEvents 

### PR DESCRIPTION
**Pull request by SmitaRKulkarni**
Changelog category (leave one):
* Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Record server startup time in ProfileEvents resolves #43188
Implementation:
* Added ProfileEvents::ServerStartupMilliseconds.
* Recorded time from start of main till listening to sockets.
Testing:
* Added a test 02532_profileevents_server_startup_time.sql

### CODE REVIEW

File: programs/server/Server.cpp [hunk 1]
```diff
@@ -1427,6 +1430,9 @@ try
 
     }
 
+    std::chrono::milliseconds startup_duration = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now() - startup_start_time);
+    ProfileEvents::increment(ProfileEvents::ServerStartupTime, startup_duration.count());
```
Comment (kssenii) [inline programs/server/Server.cpp:1434]: It is not the right place, at this moment startup is not finished, AFAIK `servers_to_start_before_tables` are keeper servers, not the clickhouse server, need to put this lower - near line 1820.

File: programs/server/Server.cpp [hunk 2]
```diff
@@ -652,6 +653,8 @@ static void sanityChecks(Server & server)
 int Server::main(const std::vector<std::string> & /*args*/)
 try
 {
+    std::chrono::system_clock::time_point startup_start_time = std::chrono::system_clock::now();
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply programs/server/Server.cpp]: addressed it

File: src/Common/ProfileEvents.cpp [hunk 3]
```diff
@@ -449,7 +449,8 @@ The server successfully detected this situation and will download merged part fr
     M(OverflowBreak, "Number of times, data processing was cancelled by query complexity limitation with setting '*_overflow_mode' = 'break' and the result is incomplete.") \
     M(OverflowThrow, "Number of times, data processing was cancelled by query complexity limitation with setting '*_overflow_mode' = 'throw' and exception was thrown.") \
     M(OverflowAny, "Number of times approximate GROUP BY was in effect: when aggregation was performed only on top of first 'max_rows_to_group_by' unique keys and other keys were ignored due to 'group_by_overflow_mode' = 'any'.") \
-
+    \
+    M(ServerStartupTime, "Time elapsed from starting server to listening to sockets")\
```
[the remark this answers is not in the document]
Comment (SmitaRKulkarni) [reply src/Common/ProfileEvents.cpp]: addressed it

File: programs/server/Server.cpp (hunk 1 above)
Comment (SmitaRKulkarni) [reply programs/server/Server.cpp]: addressed it, thanks

File: programs/server/Server.cpp [hunk 4]
```diff
@@ -1810,6 +1813,8 @@ try
                                                                      &CurrentMetrics::MaxDDLEntryID, &CurrentMetrics::MaxPushedDDLEntryID));
         }
 
+        ProfileEvents::increment(ProfileEvents::ServerStartupMilliseconds, startup_watch.elapsedMilliseconds());
```
Comment (kssenii) [inline programs/server/Server.cpp:1816]: may just a little lower - after these lines?
https://github.com/ClickHouse/ClickHouse/blob/a0fe26f5069dd425f3862a3eec4cca6066c7a7d5/programs/server/Server.cpp#L1826-L1828

Comment (kssenii) [inline programs/server/Server.cpp:1816]:
```suggestion
        startup_watch.stop()
        ProfileEvents::increment(ProfileEvents::ServerStartupMilliseconds, startup_watch.elapsedMilliseconds());
```

**Comment by qoega** [comment]
Do you know if internal port is up before we accept external http/native connections? I had an integration test where it was important to check it is up.
Don't you think it will be useful to add Events per interface as a follow up?

**Comment by alexey-milovidov** [comment]
It's ok to have a single event for the time when all the listeners are up.

**Comment by qoega** [comment]
Still, do we guarantee that all listeners are up when we write those? What is the order of endpoint initialization? How can you check that all are ready - just OK on http or native will not work and query event is inadequate.

**Comment by alexey-milovidov** [comment]
They are initialized in a sequence and are ready at similar times (except the embedded Keeper, but it does not matter). It's ok if the metric overestimates the startup time a little. Therefore, let's record the time before starting the last of the listeners.

File: programs/server/Server.cpp (hunk 4 above)
Comment (SmitaRKulkarni) [reply programs/server/Server.cpp]: addressed it

Comment (SmitaRKulkarni) [reply programs/server/Server.cpp]: addressed it

**Review by kssenii — APPROVED** [review]

**Pull request merged**